pkg/auth/auth.go

```diff
@@ -0,0 +1,250 @@
+
+		if c.AuthConfig.UseUserIdentity {
+			if len(c.AuthConfig.UserIdentityClientId) == 0 {
+				return fmt.Errorf("neither AZURE_CLIENT_ID or VIRTUALNODE_USER_IDENTITY_CLIENTID is being set")
+			}
+
+			log.G(context.TODO()).Info("Using user identity for Authentication")
+		}
+
+		if tenantID := os.Getenv("AZURE_TENANT_ID"); tenantID != "" {
+			c.AuthConfig.TenantID = tenantID
+		}
+
+		if subscriptionID := os.Getenv("AZURE_SUBSCRIPTION_ID"); subscriptionID != "" {
+			c.AuthConfig.SubscriptionID = subscriptionID
+		}
+	}
+
+	resource := c.Cloud.Services[cloud.ResourceManager].Endpoint
+
+	c.Authorizer, err = c.getAuthorizer(resource)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Authentication represents the Authentication file for Azure.
+type Authentication struct {
+	ClientID             string `json:"clientId,omitempty"`
+	ClientSecret         string `json:"clientSecret,omitempty"`
+	SubscriptionID       string `json:"subscriptionId,omitempty"`
+	TenantID             string `json:"tenantId,omitempty"`
+	UseUserIdentity      bool   `json:"useUserIdentity,omitempty"`
+	UserIdentityClientId string `json:"userIdentityClientId,omitempty"`
+}
+
+// NewAuthenticationFromFile returns an Authentication struct from file path
+func (a *Authentication) NewAuthenticationFromFile(filepath string) error {
+	b, err := ioutil.ReadFile(filepath)
+	if err != nil {
+		return fmt.Errorf("reading Authentication file %q failed: %v", filepath, err)
+	}
+
+	// Authentication file might be encoded.
+	decoded, err := a.decode(b)
+	if err != nil {
+		return fmt.Errorf("decoding Authentication file %q failed: %v", filepath, err)
+	}
+
+	// Unmarshal the Authentication file.
+	if err := json.Unmarshal(decoded, &a); err != nil {
+		return err
+	}
+	return nil
+}
+
+// NewAuthentication returns an Authentication struct from user provided credentials.
+func NewAuthentication(clientID, clientSecret, subscriptionID, tenantID, userAssignedIdentityID string) *Authentication {
+
+	return &Authentication{
+		ClientID:             clientID,
+		ClientSecret:         clientSecret,
+		SubscriptionID:       subscriptionID,
+		TenantID:             tenantID,
+		UserIdentityClientId: userAssignedIdentityID,
+	}
+}
+
+// aksCredential represents the credential file for AKS
+type aksCredential struct {
+	Cloud                  string `json:"cloud"`
+	TenantID               string `json:"tenantId"`
+	SubscriptionID         string `json:"subscriptionId"`
+	ClientID               string `json:"aadClientId"`
+	ClientSecret           string `json:"aadClientSecret"`
+	ResourceGroup          string `json:"resourceGroup"`
+	Region                 string `json:"location"`
+	VNetName               string `json:"vnetName"`
+	VNetResourceGroup      string `json:"vnetResourceGroup"`
+	UserAssignedIdentityID string `json:"userAssignedIdentityID"`
+}
+
+// NewAKSCredential returns an aksCredential struct from file path
+func NewAKSCredential(p string) (*aksCredential, error) {
```

Comment:
Can this be an internal method?  rename p-> filepath?

Reply:
done